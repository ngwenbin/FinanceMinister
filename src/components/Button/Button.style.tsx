import { StyleSheet } from "react-native";
import { Colors } from "@constants/Colors";
import { getFont, normalize } from "@utils";

export const buttonStyles = StyleSheet.create({
  buttonContainer: {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
  },
  xs: {
    paddingHorizontal: normalize(11),
    paddingVertical: normalize(7),
    borderRadius: normalize(6),
    _flexGap: normalize(4),
    ...getFont("medium", "xs"),
  },
  sm: {
    paddingHorizontal: normalize(16),
    paddingVertical: normalize(8),
    borderRadius: normalize(6),
    _flexGap: normalize(8),
    ...getFont("medium", "sm"),
  },
  md: {
    paddingHorizontal: normalize(18),
    paddingVertical: normalize(8),
    borderRadius: normalize(8),
    _flexGap: normalize(8),
    ...getFont("medium", "md"),
  },
  lg: {
    paddingHorizontal: normalize(20),
    paddingVertical: normalize(9),
    borderRadius: normalize(8),
    _flexGap: normalize(8),
    ...getFont("medium", "lg"),
  },
  solid: {
    color: Colors.white,
    backgroundColor: Colors.primary[500],
    _before: {
      backgroundColor: Colors.primary[500],
    },
    _after: {
      backgroundColor: Colors.primary[700],
    },
    _disabled: {
      backgroundColor: Colors.primary[200],
    },
  },
  outline: {
    borderStyle: "solid",
    borderWidth: 1,
    color: Colors.primary[500],
    _before: {
      borderColor: Colors.primary[500],
    },
    _after: {
      borderColor: Colors.primary[500],
    },
    _disabled: {
      color: Colors.primary[200],
      borderColor: Colors.primary[100],
    },
  },
  white: {
    borderStyle: "solid",
    borderWidth: 1,
    color: Colors.gray[700],
    _before: {
      borderColor: Colors.gray[300],
    },
    _after: {
      borderColor: Colors.gray[300],
    },
    _disabled: {
      color: Colors.gray[400],
      backgroundColor: Colors.gray[100],
      borderColor: Colors.gray[100],
    },
  },
});
