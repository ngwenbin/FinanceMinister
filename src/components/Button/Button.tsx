import { Text, Pressable, View, ViewStyle } from "react-native";
import { buttonStyles as styles } from "./Button.style";

type ButtonSizes = "xs" | "sm" | "md" | "lg";
type ButtonVariant = "solid" | "outline" | "white";

interface ButtonProps {
  adornment?: JSX.Element;
  adornmentPosition?: "left" | "right";
  containerStyle?: ViewStyle;
  label?: string;
  size?: ButtonSizes;
  variant?: ButtonVariant;
  onPressCallback?: () => void;
  onLongPressCallback?: () => void;
}

const Button = ({
  adornment,
  adornmentPosition,
  containerStyle,
  label,
  size = "sm",
  variant = "solid",
  onPressCallback,
  onLongPressCallback,
}: ButtonProps) => {
  return (
    <Pressable
      style={({ pressed }) => [
        styles.buttonContainer,
        styles[size],
        styles[variant],
        { flexDirection: adornmentPosition === "left" ? "row-reverse" : "row" },
        { ...(pressed ? styles[variant]._after : styles[variant]._before) },
        containerStyle,
      ]}
      onPress={onPressCallback}
      onLongPress={onLongPressCallback}
    >
      {label ? (
        <Text
          style={{
            color: styles[variant].color,
            fontFamily: styles[size].fontFamily,
            fontSize: styles[size].fontSize,
          }}
        >
          {label}
        </Text>
      ) : null}
      {adornment ? (
        <View
          style={{
            [adornmentPosition === "left" ? "paddingRight" : "paddingLeft"]:
              styles[size]._flexGap,
          }}
        >
          {adornment}
        </View>
      ) : null}
    </Pressable>
  );
};

export default Button;
