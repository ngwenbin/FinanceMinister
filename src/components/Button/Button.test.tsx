import "@testing-library/jest-native/extend-expect";
import { render, screen } from "@testing-library/react-native";
import { testIds } from "@utils/tests/testIds";
import { Text } from "react-native";
import Button from "./Button";

const buttonTitle = "Test button";
const buttonTitleTestId = testIds.sharedComponents.button.buttonTitle;
const buttonAdornmentTestId =
  testIds.sharedComponents.button.buttonAdornmentContainer;

describe("<Button>", () => {
  describe("as default", () => {
    // const onPressCallback = jest.fn(() => null);
    // const onLongPressCallback = jest.fn(() => null);

    test("Renders successfully", () => {
      const activeButton = render(<Button title={buttonTitle} />);
      expect(activeButton).toBeDefined();
    });

    test("Renders correctly", () => {
      const activeButton = render(<Button title={buttonTitle} />);
      expect(activeButton.toJSON()).toMatchSnapshot();
    });

    test("Outputs the correct text", () => {
      render(<Button title={buttonTitle} />);
      expect(screen.getByTestId(buttonTitleTestId)).toHaveTextContent(
        buttonTitle
      );
    });

    test(`Does not render an adornment`, () => {
      render(<Button title={buttonTitle} />);
      expect(screen.queryByTestId(buttonAdornmentTestId)).toBeNull();
    });

    test(`Renders an adornment`, () => {
      render(<Button title={buttonTitle} adornment={<Text>Hello</Text>} />);
      expect(screen.queryByTestId(buttonAdornmentTestId)).toBeDefined();
    });
  });
});
