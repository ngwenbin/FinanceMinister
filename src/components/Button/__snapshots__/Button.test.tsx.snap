// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> as default Renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "center",
      },
      Object {
        "_flexGap": 10,
        "borderRadius": 6,
        "fontFamily": "Inter_500Medium",
        "fontSize": 14,
        "paddingHorizontal": 20,
        "paddingVertical": 10,
      },
      Object {
        "_after": Object {
          "backgroundColor": "#022CB3",
        },
        "_before": Object {
          "backgroundColor": "#054EFA",
        },
        "_disabled": Object {
          "backgroundColor": "#9AC1FE",
        },
        "backgroundColor": "#054EFA",
        "color": "#FFFFFF",
      },
      Object {
        "flexDirection": "row",
      },
      Object {
        "backgroundColor": "#054EFA",
      },
      Object {},
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Inter_500Medium",
          "fontSize": 14,
        },
        Object {},
      ]
    }
    testID="button-0"
  >
    Test button
  </Text>
</View>
`;
